shared/lib/src/bloc_complex/catalog/catalog_bloc.dart

```diff
@@ -0,0 +1,135 @@
+import 'dart:async';
+import 'dart:math';
+
+import 'package:flutter/widgets.dart';
+import 'package:reactive_exploration/common/models/product.dart';
+import 'package:reactive_exploration/src/bloc_complex/catalog/catalog_page.dart';
+import 'package:reactive_exploration/src/bloc_complex/catalog/catalog_slice.dart';
+import 'package:rxdart/rxdart.dart';
+
+/// This component encapsulates the logic of fetching products from
+/// a database, page by page, according to position in an infinite list.
+///
+/// Only the data that are close to the current location are cached, the rest
+/// is thrown away.
+class CatalogBloc {
+  static const _productsPerPage = 10;
+
+  /// We're using ReactiveX's [PublishSubject] here because we want to easily
+  /// buffer the stream. See [CatalogBloc] constructor.
+  final _indexController = PublishSubject<int>();
```

Comment:
nit: There's some inconsistency in that not all the fields are explicitly typed [though most are] - it would be nice if this was declared as `final PublishSubject<int> _indexController` for example.

[Apologies for the drive-by comment - this repo is great!]

Reply:
Thanks! Good catch. I just made the types in the PR consistent according to [Effective Dart](https://www.dartlang.org/guides/language/effective-dart/design#types) with https://github.com/filiph/state_experiments/pull/2/commits/b427d28e980974e252ebb3bf466b587465ab61c7.

Later I'll make the types consistent everywhere.

In the case of this particular line, this is in line with Effective Dart. Type here is inferred, and on top of that the member is private, so no need to annotate.